plane/src/typed_unix_socket/client.rs

```diff
@@ -0,0 +1,247 @@
+        let (response_tx, response_rx) = oneshot::channel();
+        let id = random_token();
+
+        let wrapper = WrappedMessage {
+            id: Some(id.clone()),
+            message: request,
+        };
+
+        self.response_map.insert(id, response_tx);
+
+        // Wait until there is at least one subscriber
+        let mut backoff = get_quick_backoff();
+        while self.tx.receiver_count() == 0 {
+            tracing::info!("Waiting for a subscriber...");
+            backoff.wait().await;
+        }
+
+        self.tx.send(wrapper)?;
+        let response = response_rx.await?;
+        Ok(response)
+    }
+
+    pub async fn send_message(&self, message: MessageToServer) -> Result<(), Error> {
+        let wrapper = WrappedMessage { id: None, message };
+
+        // Wait until there is at least one subscriber
+        let mut backoff = get_quick_backoff();
+        while self.tx.receiver_count() == 0 {
+            tracing::info!("Waiting for a subscriber...");
+            backoff.wait().await;
+        }
+
+        self.tx.send(wrapper)?;
+        Ok(())
+    }
+
+    pub fn subscribe_events(&self) -> broadcast::Receiver<MessageToClient> {
+        self.event_tx.subscribe()
+    }
+}
+
+impl<MessageToServer, MessageToClient> Drop
+    for TypedUnixSocketClient<MessageToServer, MessageToClient>
+where
+    MessageToServer: Send + Sync + 'static + Clone + Debug + Serialize + for<'de> Deserialize<'de>,
+    MessageToClient: Send + Sync + 'static + Clone + Debug + Serialize + for<'de> Deserialize<'de>,
+{
+    fn drop(&mut self) {
+        let _ = self.shutdown_tx.send(());
+    }
+}
+
+async fn handle_connection<MessageToServer, MessageToClient>(
+    stream: UnixStream,
+    mut rx: broadcast::Receiver<WrappedMessage<MessageToServer>>,
+    response_map: Arc<DashMap<String, oneshot::Sender<MessageToClient>>>,
+    event_tx: broadcast::Sender<MessageToClient>,
+    shutdown_rx: watch::Receiver<()>,
+) -> Result<(), Box<dyn std::error::Error + Send + Sync>>
+where
+    MessageToServer: Send + Sync + 'static + Clone + Debug + Serialize + for<'de> Deserialize<'de>,
+    MessageToClient: Send + Sync + 'static + Clone + Debug + Serialize + for<'de> Deserialize<'de>,
+{
+    let (read_half, write_half) = tokio::io::split(stream);
+    let reader = BufReader::new(read_half);
+    let writer = BufWriter::new(write_half);
+
+    let mut lines = reader.lines();
+    let mut writer = writer;
+
+    // Task to handle receiving messages
+    let recv_task = tokio::spawn({
+        let event_tx = event_tx.clone();
+        let response_map = Arc::clone(&response_map);
+        let mut shutdown_rx = shutdown_rx.clone();
+        async move {
+            loop {
+                tokio::select! {
+                    _ = shutdown_rx.changed() => {
+                        tracing::info!("Shutting down receive task");
+                        break;
+                    }
+                    result = lines.next_line() => {
```

Comment:
Should maybe have a timeout? Feel free to ignore if y'all have already considered this and decided against it.